library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,76 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else calling the function
+/// via that function pointer is Undefined Behavior. ABI compatibility is a lot stricter than merely
+/// having the same memory layout; for example, even if `i32` and `f32` have the same size and
+/// alignment, they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>` (specifically, only `Box<T, Global>`),
+///   `NonNull<T>` are all ABI-compatible with each other for all `T`. Two of these pointer types
+///   with different `T` are ABI-compatible if they have the same metadata type (`<T as
+///   Pointee>::Metadata`).
+/// - `usize` is ABI-compatible with the `uN` integer type of the same size, and likewise `isize` is
+///   ABI-compatible with the `iN` integer type of the same size.
+/// - Any two `fn`/`extern "ABI" fn` types are ABI-compatible with each other, independent of their
```

Comment:
This sounds a bit strong. Some function pointer types, especially those with bespoke ABIs, might not be "real" function pointers or live in a different address space or the like. I'm not sure we should close this door with an overly broad comment like this.

Reply:
AIUI this is currently intended to only make function pointers with the same ABI string guaranteed to be ABI compatible, so that shouldn't matter; while between ABIs function pointers could look different, I have a hard time imagining an ABI that passes function pointers differently based on that function's signature.

If that's the intent, I'd explicitly state "with the same ABI string" here, because as-is it's unclear whether this is meant to require the ABI string to match or to ignore the ABI string.

Reply:
Oh yes that's a good point. If that's what was meant then I indeed misread the statement.

Reply:
No this is intended to make all `fn` ptrs compatible even with different ABI strings. Looks like I should change that.

However, "same ABI string" is too restrictive. I'm fairly sure we want `extern "C" fn` and `extern "C-unwind" fn` to be ABI-compatible. So it should be about ABI-compatible ABI strings. This is getting confusing. ;)